docs/weave-tutorial/03-codec.md

```diff
@@ -0,0 +1,219 @@
+
+## Create Codec file
+
+First imagine the shape of your classes. The product of your imagination will be called `Codec`. And the language of expressing shape of the classes is: [proto3](https://developers.google.com/protocol-buffers/docs/proto3). There are several different `int encodings`, `byte slices`, `strings`, and nested structures. And fields may be repeated. So forget complex types with methods now and just focus on the actual data structure.
+
+Codec is the first component that needs to be designed. Keep in mind that this part is the most important since the whole module will depend on _codec_. Codec resembles  _model_ in mvc pattern. Yet codec not only defines the model representation but marshalling/unmarshalling and interactions with messages. Codec defines the whole application state models and more. On this article, we would not go as far as explaining every proto message. One example for each message and state would be enough. Furthermore, if you want to see all the implementation, you can check out [tutorial](https://github.com/iov-one/tutorial/blob/master/x/orderbook/codec.proto)
+
+## State
+
+In the previous section, we explained how to define your domain thus attributes if models. Here `Trade` is defined as proto message. The [x/codec.proto](https://github.com/iov-one/tutorial/blob/master/x/orderbook/codec.proto#L75-L88) file defines the `Trade` type rather simply, once you remove the comments, this is all that is left:
+
+```protobuf
+// Trade is a settled partial/full order
+// We store these as independent entities to help with queries to map
+// the prices over time. They are also referenced by the Orders, so we can
+// see how much was fulfilled.
+//
+// Ask and Bid tickers are "inherited" from the orderbook
+message Trade {
+  weave.Metadata metadata = 1;
+  bytes id = 2 [(gogoproto.customname) = "ID"];
+  bytes order_book_id = 3 [(gogoproto.customname) = "OrderBookID"];
+  bytes order_id = 4 [(gogoproto.customname) = "OrderID"];
+  // Address of taker (this is an order that was instantly fulfilled)
+  bytes taker = 5 [(gogoproto.casttype) = "github.com/iov-one/weave.Address"];
+  // Address of maker (this is an order that was stored first before fulfillment)
+  bytes maker = 6 [(gogoproto.casttype) = "github.com/iov-one/weave.Address"];
+  // this is how much each side paid (they got the opposite one)
+  coin.Coin maker_paid = 7;
+  coin.Coin taker_paid = 8;
+  // executed_at defines execution time of an order
+  int64 executed_at = 9 [(gogoproto.casttype) = "github.com/iov-one/weave.UnixTime"];
+}
+```
+
+As you can see above, weave is heavily using `bytes` for identities, addresses, etc. At first glance, this might seem difficult to handle but if you take a look at [x/orderbook/bucket](https://github.com/iov-one/tutorial/blob/master/x/orderbook/bucket.go#L125) you can see it is useful for indexing and performance
```

Comment:
```suggestion
As you can see above, weave is heavily using `bytes` for identities, addresses, etc. At first glance, this might seem difficult to handle but if you take a look at [x/orderbook/bucket](https://github.com/iov-one/tutorial/blob/master/x/orderbook/bucket.go#L125) you can see it is useful for indexing and performance.
```